### ISSUE TITLE
Research needed API endpoints and create concept

### ISSUE THREAD

**Issue by patrickcerny** [AlpsBTE/alpsbte-website#14]
Resarch / ask @Cinnazeyy  or @LordTuxn  what the api needs to have / what endpoints are needed. If they dont know what you mean by that, just ask them what data needs to be send from the client to the server and back (for example: applications will need to be send to the database; image uploading is needed; etc). Just find out what exactly the API should do.

**Comment by Georg-code** [comment]
### API Concept:

**GET**
- Images
- Text
- Metadata

_Text_route_specific:_ (extra bc of lazy loading)
- Structure: /api/web/public/text/{route}/
- Returns: JSON

_Text_gobal:_
- Structure: /api/web/public/text/global/
- Returns: JSON

_Images:_
- Structure: /api/web/image/{imageId}
- Returns: Image (*.webp)

_Metadata:_ (Navbar Items, Footer Items etc)
- Structure: /api/web/public/meta
- Returns: JSON


!PLANS!
- Applications should be done through the website and API
- Every text should be editable (maybe implement there a review process?)
- Every Image should be replaceable
- Number of people with the "Builder" rank on the minecraft server
- Auth (Explained below)
- (Web Stats System? -> UserAgent, Device, Country, etc to find out the primarly use cases of the website) 

// unlikely - Very Spontaneous Idea: Implement some Plan (Player Analytics) stats somewhere: ex: On the last 30days xyz [Only has a directJava API]

For the moment we don't need any POST
As soon as we talk about the auth and "logged in user" * part we can create concepts for that.
There aren't any strict plans for that part. Someone told me, that there will be a official bte API for applications: I'm not quite sure how this api will look like. I suggest to wait until thats finished

* We yet talked about discord OAuth2 cause we can implement the discord role based system directly to the webseite. + Everyone here has a DiscordAcc + We don't have to deal with sensible data

If I can think about more, I will add that ofc. I still have to talk to @Cinnazeyy or @LordTuxn

**Comment by patrickcerny** [comment]
Very well done! That's a great starting point, should we close the issue or are there any more comments comming?

**Comment by b3h3m0th** [comment]
What are the thouhgts behind endpoint names like `/api/web/public/text` or `/api/web/image`?

**Comment by patrickcerny** [comment]
@b3h3m0th these are returning given images for the website, so other people wil be able to change these easier. There will be an admin panel, so thats why we also want the text to be responsive etc.

**Comment by b3h3m0th** [comment]
Obviously. But why `/web` and `/public`? Kinda weird naming I think. I was asking for the reason for that.

**Comment by Georg-code** [comment]
It simply has to do with the fact that we plan to create sections later that are not public. Then it will be clearer if there are some that require a key.
And web is there in case we want to use the api for sections outside the website as well. (Ideas are there: certain plugins, Discord Bot, etc)

**Comment by b3h3m0th** [comment]
**tl;dr**
Not a fan of it. The API shouldn't care WHO (in terms of whether it's the website, the bot etc.) requests data. Name routes/sections after the corresponding content it delivers, nothing else. If a requestor doesn't have permission just return the standard forbidden error. 

Ah alright, didn't know we were gonna use the API for things other than the website. However, is it really a good idea naming endpoints after certain creterion and separating them into that kind of usage sections? Personally I believe an endpoint should just be named after the content it provides. Because...

- What if an endpoint is used by more than just one service? E.g. website AND bot AND third parties

- What if an endpoint is only available for specific groups with specific perms? More concrete than just private/public. For instance *admin* only or something similar... 

- What if we decide a route suddenly becomes private or the other way around? You would have to adjust a fucking lot in the code (regarding imports to the other access section) and probably in the file structure as well which is going to be really annoying)


- Besides, this architecture would make a lot of things really confusing. You always would have to look up first whether a route is defined in which "sections".

**Comment by Georg-code** [comment]
Hmm yes I see. I haven't looked at it from the side yet. The best thing is to discuss how we can best set it up in the meeting tomorrow.
